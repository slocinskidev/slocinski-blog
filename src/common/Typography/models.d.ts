import { ElementType } from 'react';

export type TypographyVariant =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'sub-h1'
  | 'sub-h2'
  | 'body1'
  | 'body2'
  | 'body3'
  | 'body4'
  | 'body5'
  | 'body6';

export type TypographyGutter = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | number;
export type TypographyAlign = 'inherit' | 'left' | 'center';
export type TypographyLineHeight = 'big' | 'regular' | 'small' | string;

export interface TypographyProps {
  variant: TypographyVariant;
  align?: TypographyAlign;
  component?: ElementType;
  customClass?: string;
  gutterBottom?: TypographyGutter;
  gutterTop?: TypographyGutter;
  gutterLeft?: TypographyGutter;
  gutterRight?: TypographyGutter;
  lineHeight?: TypographyLineHeight;
  bold?: boolean;
  black?: boolean;
  italic?: boolean;
  children?: React.ReactNode;
  dangerouslySetInnerHTML?: string;
}
