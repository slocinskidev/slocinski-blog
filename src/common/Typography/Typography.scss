@import 'styles/common/component-set.scss';
@import './utils.scss';

.typography {
  font-weight: auto;

  &--bold {
    font-weight: 600;
  }

  &--black {
    font-weight: 900;
  }

  &--italic {
    font-style: italic;
  }

  &--h1 {
    @include font-size(26);

    @include breakpoint(md-up) {
      @include font-size(57);
    }
  }

  &--h2 {
    @include font-size(24);

    @include breakpoint(md-up) {
      @include font-size(42);
    }
  }

  &--h3 {
    @include font-size(24);

    @include breakpoint(md-up) {
      @include font-size(32);
    }
  }

  &--sub-h1 {
    @extend .typography--h2;
  }

  &--sub-h2 {
    @include font-size(18);

    @include breakpoint(md-up) {
      @include font-size(24);
    }
  }

  &--body1 {
    @include font-size(16);

    @include breakpoint(xs-up) {
      @include font-size(20);
    }
  }

  &--body2 {
    @include font-size(14);

    @include breakpoint(md-up) {
      @include font-size(18);
    }
  }

  &--body3 {
    @include font-size(14);
  }

  &--body4 {
    @include font-size(14);

    @include breakpoint(md-up) {
      @include font-size(16);
    }
  }

  &--body5 {
    @include font-size(12);
  }

  &--body6 {
    @include font-size(22);
  }
}

@include lineHeight('big');
@include lineHeight('regular');
@include lineHeight('small');

@include gutter('bottom');
@include gutter('top');
@include gutter('left');
@include gutter('right');

@include align('inherit');
@include align('left');
@include align('center');
