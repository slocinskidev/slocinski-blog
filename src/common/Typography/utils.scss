// Variables
$lineHeightList: (
  big: 2,
  regular: 1.5,
  small: 0.9,
);

$defaultPaddings: (
  1: $gap-xxsmall,
  2: $gap-small,
  3: $gap-regular,
  4: $gap-large,
  5: $gap-xlarge,
  6: $gap-xxlarge,
  7: $gap-xxxlarge,
  8: $gap-huge,
);

// Mixins
@mixin lineHeight($size) {
  .typography--line-height-#{$size} {
    line-height: map-get($lineHeightList, $size);
  }
}

@mixin gutter($side) {
  @for $i from 1 through length($defaultPaddings) {
    .typography--gutter-#{$side}-#{$i} {
      padding-#{$side}: map-get($defaultPaddings, $i);
    }
  }
}

@mixin align($option) {
  .typography--#{$option} {
    text-align: #{$option};
  }
}

@mixin font-size($target, $context: $font-size-base) {
  font-size: ($target / $context) * 1rem;
}
