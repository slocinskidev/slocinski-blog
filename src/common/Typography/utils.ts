import { JSXTags } from './constants';

export const variantToComponentMapping = {
  h1: JSXTags.H1,
  h2: JSXTags.H2,
  h3: JSXTags.H3,
  'sub-h1': JSXTags.H1,
  'sub-h2': JSXTags.H2,
  body1: JSXTags.PARAGRAPH,
  body2: JSXTags.PARAGRAPH,
  body3: JSXTags.PARAGRAPH,
  body4: JSXTags.PARAGRAPH,
  body5: JSXTags.PARAGRAPH,
  body6: JSXTags.PARAGRAPH,
};
