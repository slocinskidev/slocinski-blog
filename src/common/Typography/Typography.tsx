import React, { FC } from 'react';
import classNames from 'classnames';

import { DangerouslySetInnerHtml } from 'layout';

import { TypographyProps } from './models';
import { variantToComponentMapping } from './utils';

import './Typography.scss';

const Typography: FC<TypographyProps> = ({
  component,
  customClass,
  variant,
  align,
  gutterBottom,
  gutterTop,
  gutterLeft,
  gutterRight,
  lineHeight,
  bold,
  black,
  italic,
  dangerouslySetInnerHTML,
  children,
}) => {
  const Component = component || variantToComponentMapping[variant];

  const componentClasses = classNames('typography', customClass, {
    [`typography--${variant}`]: variant,
    [`typography--gutter-bottom-${gutterBottom}`]: gutterBottom,
    [`typography--gutter-top-${gutterTop}`]: gutterTop,
    [`typography--gutter-left-${gutterLeft}`]: gutterLeft,
    [`typography--gutter-right-${gutterRight}`]: gutterRight,
    [`typography--line-height-${lineHeight}`]: lineHeight,
    [`typography--${align}`]: align,
    'typography--bold': bold,
    'typography--black': black,
    'typography--italic': italic,
  });

  const render = dangerouslySetInnerHTML ? (
    <DangerouslySetInnerHtml
      element={component}
      data-testid="typography"
      className={componentClasses}
      html={dangerouslySetInnerHTML}
    />
  ) : (
    <Component data-testid="typography" className={componentClasses}>
      {children}
    </Component>
  );

  return render;
};

export default Typography;
