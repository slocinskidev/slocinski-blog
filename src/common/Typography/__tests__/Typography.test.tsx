import React from 'react';
import { render, screen } from 'testUtils';

import {
  bodyVariantClassname,
  TEST_TXT,
  typographyClassesMock,
  TypographyPropsMock,
} from '../__mock__/mock';
import { TypographyProps } from '../models';
import Typography from '../Typography';

const defaultProps: TypographyProps = {
  ...(TypographyPropsMock as TypographyProps),
};

it('Typography should render correcty with required and optional props', () => {
  const { rerender } = render(<Typography {...defaultProps}>{TEST_TXT}</Typography>);

  let TypographyNode = screen.getByTestId('typography');

  // it renders properly with all optional props
  expect(TypographyNode).toBeInTheDocument();

  // Create snapshot for CR purpouses
  expect(TypographyNode).toMatchSnapshot();

  // TypographyNode renders with all optional classnames received from props
  typographyClassesMock.forEach((classname) => {
    expect(TypographyNode).toHaveClass(classname);
  });

  rerender(<Typography variant="body1">{TEST_TXT}</Typography>);

  TypographyNode = screen.getByTestId('typography');

  // it renders without optional classnames when not provided
  typographyClassesMock.forEach((classname) => {
    expect(TypographyNode).not.toHaveClass(classname);
  });

  // it should not be present as heading when body1 variant is provided
  expect(screen.queryByRole('heading')).toBeNull();

  // it should receive body1 classname when body1 variant provided
  expect(TypographyNode).toHaveClass(bodyVariantClassname);
});
