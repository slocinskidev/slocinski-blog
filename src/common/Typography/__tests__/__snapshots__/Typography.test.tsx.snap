// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography should render correcty with required and optional props 1`] = `
<h1
  class="typography testClass typography--h1 typography--gutter-bottom-1 typography--gutter-top-1 typography--gutter-left-1 typography--gutter-right-1 typography--line-height-big typography--center typography--bold typography--italic"
  data-testid="typography"
>
  Test Lorem Ipsum
</h1>
`;
